packages/core/upload/server/controllers/admin-folder-file.js

```diff
@@ -60,25 +61,136 @@ module.exports = {
+    try {
+      // fetch folders
+      const existingFolders = await strapi.db
+        .queryBuilder(FOLDER_MODEL_UID)
+        .select(['id', 'uid', 'path'])
+        .where({ id: { $in: folderIds } })
+        .transacting(trx)
+        .forUpdate()
+        .execute();
+
+      // fetch files
+      const existingFiles = await strapi.db
+        .queryBuilder(FILE_MODEL_UID)
+        .select(['id'])
+        .where({ id: { $in: fileIds } })
+        .transacting(trx)
+        .forUpdate()
+        .execute();
+
+      // fetch destinationFolder path
+      let destinationFolderPath = '/';
+      if (destinationFolderId !== null) {
+        const destinationFolder = await strapi.db
+          .queryBuilder(FOLDER_MODEL_UID)
+          .select('path')
+          .where({ id: destinationFolderId })
+          .transacting(trx)
+          .first()
+          .execute();
+        destinationFolderPath = destinationFolder.path;
+      }
+
+      const fileTable = strapi.getModel(FILE_MODEL_UID).collectionName;
+      const folderTable = strapi.getModel(FOLDER_MODEL_UID).collectionName;
+      const folderPathColName = strapi.db.metadata.get(FILE_MODEL_UID).attributes.folderPath
+        .columnName;
+      const pathColName = strapi.db.metadata.get(FOLDER_MODEL_UID).attributes.path.columnName;
+
+      if (existingFolders.length > 0) {
+        // update folders' parent relation (delete + insert; upsert not possible)
+        const joinTable = strapi.db.metadata.get(FOLDER_MODEL_UID).attributes.parent.joinTable;
+        await strapi.db
+          .queryBuilder(joinTable.name)
+          .transacting(trx)
+          .delete()
+          .where({ [joinTable.joinColumn.name]: { $in: folderIds } })
+          .execute();
+        await strapi.db
+          .queryBuilder(joinTable.name)
+          .transacting(trx)
+          .insert(
+            existingFolders.map(folder => ({
+              [joinTable.inverseJoinColumn.name]: destinationFolderId,
+              [joinTable.joinColumn.name]: folder.id,
+            }))
+          )
+          .execute();
 
-    const updatedFolders = [];
-    // updates are done in order (not in parallele) to avoid mixing queries (path)
-    for (let folderId of folderIds) {
-      const updatedFolder = await folderService.update(
-        folderId,
-        { parent: destinationFolderId },
-        { user }
-      );
-      updatedFolders.push(updatedFolder);
+        for (const existingFolder of existingFolders) {
+          // update path for folders themselves & folders below
+          await strapi.db
+            .connection(folderTable)
+            .transacting(trx)
+            .where(pathColName, 'like', `${existingFolder.path}%`)
+            .update(
+              pathColName,
+              strapi.db.connection.raw('REPLACE(??, ?, ?)', [
+                pathColName,
+                existingFolder.path,
+                joinBy('/', destinationFolderPath, existingFolder.uid),
```

Comment:
Using replace can break the paths here because it replaces every occurrence it finds and you cannot be sure the folder structure is not repeating somehow.

E.g you have a path like that `/my-folder/tinny-folder/my-folder/my-uid` and move the first `my-folder` to `new-folder`, the replace will actually rename the second `my-folder` too.

In this situation I think you could look into using substring to only keep the right part and only append the new folder path

Reply:
Yes I thought about that too. Here was my thinking: 
- the path is using uid v4. Shouldn't we rely on the fact that each `uid` is unique?
- if we assume the uid is unique, then the path can repeat itself only if a folder contains itself. It means the folder structure is already broken.

Do you still think it's valuable to use substring (I can do that, this [example](https://stackoverflow.com/questions/18570965/update-substring-of-a-column) seems good)

Reply:
ok :) No breadcrumb so far. Too complex for the v1 we wanted to do